### PR TITLE
Update NixOS Module evremap

### PR DESCRIPTION
**Pull request by pixL404**
I have updated the NixOS Module evremap.
I expanded the type for remappable keys to include `BTN`.
I added the optional `phys` attribute, which can be used if multiple devices share the same name. As toml doesn't support `null` values, I added a filter function to the toml generation.

When testing the `phys` attribute, an error occurs that prevents the service to start if you try to switch to the new generation without a reboot. The software wants to create a uinput device but some permissions were not set immediately. After rebooting the same config worked. I'm unsure if this is something to include in the description of the attribute.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/misc/evremap.nix [hunk 1]
```diff
@@ -117,7 +129,9 @@ in
       description = "evremap - keyboard input remapper";
       wantedBy = [ "multi-user.target" ];
 
-      script = "${lib.getExe pkgs.evremap} remap ${format.generate "evremap.toml" cfg.settings}";
+      script = "${lib.getExe pkgs.evremap} remap ${
+        format.generate "evremap.toml" (lib.attrsets.filterAttrs (n: v: v != null) cfg.settings)
```
Comment (pluiedev) [inline nixos/modules/services/misc/evremap.nix:133]: This is getting a bit lengthy so maayyyyyybe consider moving it out

**Review by pluiedev — APPROVED** [review]
Why not

File: nixos/modules/services/misc/evremap.nix (hunk 1 above)
Comment (pixL404) [reply nixos/modules/services/misc/evremap.nix:133]: sure, how about like this:
```nix
let
  cfg = config.services.evremap;
  format = pkgs.formats.toml { };

  settings = lib.attrsets.filterAttrs (n: v: v != null) cfg.settings;

...

    script = "${lib.getExe pkgs.evremap} remap ${format.generate "evremap.toml" settings}";
```

Comment (pluiedev) [reply nixos/modules/services/misc/evremap.nix:133]: Oh I intended to mean to move the generated `evremap.toml` derivation out, and name it something like `configFile`, and letting the script where it is

Comment (pixL404) [reply nixos/modules/services/misc/evremap.nix:133]: that does make a lot more sense than my assumption you were talking about my changes only

I think this should be satisfying then?

```nix
let
  cfg = config.services.evremap;
  format = pkgs.formats.toml { };

  settings = lib.attrsets.filterAttrs (n: v: v != null) cfg.settings;
  configFile = format.generate "evremap.toml" settings;

...

  script = "${lib.getExe pkgs.evremap} remap ${configFile}";
```

I can push this if you think its fitting.

What should we do about the description of `phys` and the reboot thing?

Comment (pluiedev) [reply nixos/modules/services/misc/evremap.nix:133]: Yeah, looks good - the description for `phys` is OK as well, maybe could change the wording a bit

File: nixos/modules/services/misc/evremap.nix [hunk 2]
```diff
@@ -63,6 +63,18 @@ in
...
 
+          phys = lib.mkOption {
+            type = lib.types.nullOr lib.types.str;
+            default = null;
+            example = "usb-0000:07:00.3-2.1.1/input0";
+            description = ''
+              If you have multiple devices with the same name, you can optionally
+              specify this attribute.
+
+              If available, the value will be printed when running `evremap list-devices` with elevated permissions.
```
Comment (pluiedev) [inline nixos/modules/services/misc/evremap.nix:74]: "If available" sounds a bit odd.
```suggestion
              The physical device name to listen on.
              
              This attribute may be specified to disambiguate multiple devices with the same device name.
              The physical device names of each device can be obtained by running `evremap list-devices` with elevated permissions.
```

File: nixos/modules/services/misc/evremap.nix (hunk 1 above)
Comment (pixL404) [reply nixos/modules/services/misc/evremap.nix:133]: Great, I'll push in the evening today

**Comment by pixL404** [comment]
@pluiedev done!

**Review by pluiedev — APPROVED** [review]
LGTM except I think you should probably squash the fixup commits and split changes based on areas you've changed (e.g. add phys option and allow BTN_* keys)

**Comment by pixL404** [comment]
Sorry if this is getting tedious but you mean in to squash in 3 commits right?
1. update type
2. add phys option (including change to `toml` generation)
3. everything else

or do you mean squash into a single commit to have the PR as a single change?

**Comment by pluiedev** [comment]
No, just the first two commits (I believe the third commit, [d22eeb20ea4c7a6561682a9757c6184870ab5020](https://github.com/NixOS/nixpkgs/pull/362661/commits/d22eeb20ea4c7a6561682a9757c6184870ab5020), is only needed because of the phys option, so it should be squashed into that commit)

**Comment by pixL404** [comment]
I think I borked it

**Comment by pixL404** [comment]
What steps are necessary to get this merged?

**Comment by pluiedev** [comment]
Get the attention of a maintainer and/or linking this PR to the "PRs already reviewed" thread on the NixOS Discourse

**Comment by pluiedev** [comment]
Get the attention of a maintainer and/or linking this PR to the "PRs already reviewed" thread on the NixOS Discourse

**Review by Scrumplex — APPROVED** [review]

**Comment by bbenne10** [comment]
ugh. I'll just cherry pick by hand and open the PR. I'll update this with the PR number when I get it sorted.

**Pull request merged**